readthedocs/projects/filters.py

```diff
@@ -0,0 +1,163 @@
+import logging
+
+from django.db.models import Count, F, Max
+from django.forms.widgets import HiddenInput
+from django.utils.translation import ugettext_lazy as _
+from django_filters import CharFilter, ChoiceFilter, FilterSet, OrderingFilter
+
+log = logging.getLogger(__name__)
+
+
+class VersionSortOrderingFilter(OrderingFilter):
+
+    """
+    Version list sort ordering django_filters filter.
+
+    Django-filter is highly opionated, and the default model filters do not work
+    well with empty/null values in the filter choices. In our case, empty/null
+    values are used for a default query. So, to make this work, we will use a
+    custom filter, instead of an automated model filter.
+
+    The empty/None value is used to provide both a default value to the filter
+    (when there is no ``sort`` query param), but also provide an option that is
+    manually selectable (``?sort=relevance``). We can't do this with the default
+    filter, because result would be params like ``?sort=None``.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.extra['choices'] = [
+            ('name', _('Name')),
+            ('-name', _('Name (descending)')),
+            ('-recent', _('Most recently built')),
+            ('recent', _('Least recently built')),
+        ]
+
+    def filter(self, qs, value):
+        # This is where we use the None value for this custom filter. This
+        # doesn't work with a standard model filter.
+        if value is None:
+            value = ['relevance']
+        return qs.annotate(
```

Comment:
Don't we only want to annotate the value that we're actually filtering? Seems like this will add query overhead.

Reply:
Yeah, that's an easy optimization to make. I'll update this